Unified tutorial and risk tensor output in all wrappers

egesko opened the pull request:
This PR implements a comprehensive tutorial that touches on all risk metrics, as well as updating all wrapper outputs to be consistent (they now all output risk tensor).

IaroslavElistratov commented on notebooks/0_Unified-Tutorial.ipynb:
Also, for this tutorial, I believe we want to add @slolla as the commit author. Do you agree with that?

IaroslavElistratov replied:
Do you think we can include [this](https://github.com/aamini/introtodeeplearning/blob/master/lab3/Part1_IntroductionCapsa.ipynb) tutorial as well?

aamini replied:
I agree with [this](https://github.com/themis-ai/capsa/pull/118/files/370b41ba5e12efe5b7d5132de49e2ca91a2a08f8#r1149333238) as well. Would be great to include the Introduction notebook as the one labeled `0_*` and then the current `0_Unified-Tutorial` can be renamed to a later number (as it is more advanced) and titled something about "Faces"

aamini replied:
I agree with [this](https://github.com/themis-ai/capsa/pull/118/files/370b41ba5e12efe5b7d5132de49e2ca91a2a08f8#r1149333238) as well. Would be great to include the Introduction notebook as the one labeled 0_* and then the current 0_Unified-Tutorial can be renamed to a later number (as it is more advanced) and titled something about "Faces"

IaroslavElistratov commented on notebooks/1_Ensemble-Classification.ipynb:
Could we please replace excerpts below with something different
- capsa (a Python Library for Mitigating Uncertainty) --> [something else]
- Uncertainty in a model's predictions due to model limitations --> I think we mean model not being expressive enough to represent the data
- This new ensemble architecture is parallelizable with tensorflow --> I don't think this is correct, we're still looping over each model one by one

IaroslavElistratov commented on notebooks/2_Histogram-Classification.ipynb:
Could we please replace excerpts below with something different
- plot your training dataset into a graph --> plot your training dataset
- requires you to several parameters --> typo
- wrap your model mode --> typo

aamini commented on capsa/bias/histogram.py:
I'm not sure if this is related but currently it looks like `test_mve.py` is throwing an error in py39 environment?

egesko replied:
@reza-themisai  it looks like an issue with the docker image being generated, any chance you can look into this?

egesko replied:
> I understand this is out of the scope for the current PR, but in [92683da](https://github.com/themis-ai/capsa/commit/92683daf7d9538d8d458db888cf8a3a442a13524) it looks you're using `unittest.TestCase` for ensemble and mve, could you please do the same for dropout, histogram and the vae tests? Otherwise our tests right now look a little bit inconsistent. It's perfectly fine if we do it in a separate PR.

I don't have a good test creation framework yet for wrappers since the main thing we test for is whether we are seeing a higher risk metric value than average in certain points in a given dataset when we train using a fixed dataset. We end up having to define what those points should be beforehand, which is fairly counterintuitive. I will generate a tool for this in a separate PR.

egesko replied:
> I'm not sure if this is related but currently it looks like `test_mve.py` is throwing an error in py39 environment?

#121 addresses this.

egesko commented:
# Latest updates

Here's what this PR currently does:

- Updates all wrappers to output `RiskTensor()` with corresponding risk metric.
- Updates `get_histogram_probability()` for HistogramWrapper and HistogramVAEWrapper so that it doesn't do softmax over entire batch. Also adds small epsilon to frequencies so that we don't get NaN likelihoods.
- VAEWrapper and HistogramVAEWrapper does not return epistemic risk with extra two dimensions with 1 length anymore.
- Added `[Tutorials]` that points to capsa tutorials page in `README.md`
- Added two tutorials `Introduction` and `Faces`.

IaroslavElistratov commented on README.md:
```diff
@@ -98,3 +99,7 @@ Checkout our <a href="https://themisai.io/capsa/tutorials">tutorials</a> on some
+
+# 💡 Support
+
+Capsa currently supports Keras Sequential models, although we are looking at possible solutions for supporting wider range of models. We can't wait to share you some new details soon! 
```
Love this

IaroslavElistratov reviewed (approved):
Thank you @egesko!

aamini reviewed (approved):
Love it! Great work @egesko!! 🚀
